src/jet/mod.rs

```diff
@@ -56,14 +57,30 @@ pub trait Application: Clone + 'static {
     ///
     /// Every jet has a unique serialization _in context of its application_.
     /// Two jets of _different_ applications may share the same serialization.
-    fn decode_jet<I: Iterator<Item = u8>>(iter: &mut BitIter<I>)
-        -> Result<Self::Jet, crate::Error>;
+    fn decode_jet<I: Iterator<Item = u8>>(iter: &mut BitIter<I>) -> Result<Self::Jet, Error> {
+        let mut code = 0u64;
+
+        for len in 1..=64 {
+            let bit = iter.read_bits_be(1).ok_or(Error::EndOfStream)?;
+            code <<= 1;
+            code += bit;
+
+            if let Some(jet) = Self::Jet::decode(code, len) {
+                return Ok(jet);
+            }
```

Comment:
In a98a69d931fddae67da8cf8d588efeb8d713c9db:

I'm really skeptical of this algorithm. It's a clever way to generically decode from a prefix code without having a bunch of complicated branches, but in the case that we encounter a bad jet, it will try to just read the entire remainder of the program searching for some valid prefix.

In fact it won't be able to do this because the `code <<= 1` line above will overflow and cause a panic, but that's even worse :).

I get that the goal is to avoid having the Haskell code produce a prefix-code-decoder algorithm since that'd be very branch heavy and difficult to do without knowing Haskell. But I think we can avoid this by having some sort of `PrefixCode` structure which internally holds a search tree, and having the haskell-generated code just add codewords to this with simple one-liners like `code.add_jet(u64, usize, Jet { ... })` or something.

Hopefully we can produce such a tree at compile-time, but if not, we could hold it in the Context object... which may even give us some form of runtime switching between different Simplicity extensions.

I also think you should pull `(u64, usize)` into a `PrefixCodeWord` type because it's difficult to read code that's passing around bare tuples like this.

Reply:
The decoding code was at best a bad compromise. Part of this PR is finding ways to encode behaviour in static files. Let's find an encoding that let's us decode jets without having to brute force.

Reply:
I don't see why `code <<= 1` should overflow. It doesn't on my end if I disable the inner if statement.

Reply:
@uncomputable are you running in debug mode? If you try to read more than 64 bits into a `u64` it will overflow.

Reply:
Just closing out this discussion -- I talked to @uncomputable on the phone who politely explained that left-shifting by a single bit is not going to overflow no matter how many times in a row you do it :)